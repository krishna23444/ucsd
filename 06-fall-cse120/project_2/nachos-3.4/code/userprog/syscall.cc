#include "boundedbuffer.h"
#include "system.h"
#include "syscall.h"
#include "synch.h"

struct ThreadArgs {
  int startOfArgPointersVirtualAddress;
  int startOfArgsVirtualAddress;
  int argc;
  char **argv;
};

//
void runUserCode(int pointerToThreadArgs) {
  if (currentThread != NULL) {
    currentThread->space->InitRegisters();   // set the initial register values
    currentThread->space->RestoreState();    // load page table register

    struct ThreadArgs *threadArgumentsStruct = (ThreadArgs*)pointerToThreadArgs;
    int argPointerVirtualAddress = threadArgumentsStruct->startOfArgPointersVirtualAddress;
    int argVirtualAddress = threadArgumentsStruct->startOfArgsVirtualAddress;
    int argc = threadArgumentsStruct->argc;
    char **argv = threadArgumentsStruct->argv;

    // Since we append the arguments to main at the end of the virtual space,
    // when InitRegisters sets the stack pointer, the code itself will end up
    // corrupting our arguments. To solve this, we set the stack pointer where
    // the old stack ended, that is the end of the virtual memory before we
    // added space for arguments. This makes sure the running code won't
    // override our arguments when executing.
    currentThread->space->writeStackPointer(argPointerVirtualAddress);

    // copy arguments into newly allocated space
    for (int i = 0; i < argc; ++i) {
      DEBUG('s', "Writing string: %s\n", argv[i]);
      DEBUG('s', "Writing pointer %d to %d\n", argVirtualAddress, argPointerVirtualAddress);
      // write the pointer to the string
      machine->WriteMem(argPointerVirtualAddress, 4, argVirtualAddress);
      
      // write the current argument string
      for (unsigned int j = 0; j < strlen(argv[i])+1; ++j) {
        DEBUG('s', "'%c' -> %d\n", argv[i][j], argVirtualAddress);
        machine->WriteMem(argVirtualAddress, 1, argv[i][j]);
        argVirtualAddress++;
      }
      DEBUG('s', "\n");
  
      argPointerVirtualAddress += 4;
    }

    DEBUG('s', "Writing arguments: %d, %d\n", argc,
          threadArgumentsStruct->startOfArgPointersVirtualAddress);

    currentThread->SaveArgs(argc, threadArgumentsStruct->startOfArgPointersVirtualAddress);
    currentThread->RestoreArgs();
    
    delete threadArgumentsStruct;

    machine->Run();     // jump to the user program
    ASSERT(FALSE);      // machine->Run never returns;
                        // the address space exits by doing the syscall "exit"
  }
}

//
SpaceId Exec(char *name, int argc, char **argv, int willJoin) {
  OpenFile *executable = fileSystem->Open(name);
  AddrSpace *space;
  SpaceId addressSpaceId;
  Thread *newProcess;

  bool isJoinable = willJoin & 0x1;
  bool isPipedWriter = willJoin & 0x2;
  bool isPipedReader = willJoin & 0x4;

  if (executable == NULL) {
    printf("Unable to open file %s\n", name);
    return 0;
  }

  // create new address space
  space = new AddrSpace();

  // check if there is space available for the new process
  addressSpaceId = space->AllocateAndLoad(executable);
  if (addressSpaceId == 0) {
    DEBUG('s', "not enough memory for program\n");
    return 0;
  }

  delete executable;      // close file

  newProcess = new Thread("exec'ed thread", isJoinable);
  newProcess->space = space;
  joinAndAddressSpacesTable.SetThread(addressSpaceId, newProcess);

  int spaceForArgPointers = argc * 4;

  int startOfArgPointersVirtualAddress = space->getEndVirtualAddress();
  int startOfArgsVirtualAddress = startOfArgPointersVirtualAddress + spaceForArgPointers;

  struct ThreadArgs* threadArguments = new struct ThreadArgs();
  threadArguments->argc = argc;
  threadArguments->argv = argv;
  threadArguments->startOfArgPointersVirtualAddress = startOfArgPointersVirtualAddress;
  threadArguments->startOfArgsVirtualAddress = startOfArgsVirtualAddress;

  DEBUG('s', "Writing Args starting: %d\n", startOfArgsVirtualAddress);
  DEBUG('s', "Writing pointers starting: %d\n", startOfArgPointersVirtualAddress);
  // compute how many bytes we need to put all the arguments on the stack
  int additionalStackSpace = 0;
  // first add space for the actual array of character pointers
  additionalStackSpace += spaceForArgPointers;
  // now add space for each actual string
  for (int i = 0; i < argc; ++i) {
    additionalStackSpace += strlen(argv[i]);
    // add another byte for NULL
    additionalStackSpace++;
  }

  DEBUG('s',"endVirtualMemory before expand: %d\n", space->getEndVirtualAddress());
  // ask address space for more space for arguments
  if (space->ExpandMemory(additionalStackSpace) == 0) {
    DEBUG('s', "could not allocate more memory for arguments\n");
    return 0;
  }
  DEBUG('s',"endVirtualMemory after expand: %d\n", space->getEndVirtualAddress());

  BoundedBuffer* pipelineWaitingWriter = currentThread->getPipelineWaitingWriter();
  if (isPipedReader) {
    if (pipelineWaitingWriter != NULL) {
      newProcess->setReadPipe(pipelineWaitingWriter);
      pipelineWaitingWriter->makePipe();
      currentThread->clearPipeline();
    }
  }

  if (isPipedWriter) {
    if (pipelineWaitingWriter != NULL) {
      DEBUG('s', "creating a writer pipe when there's already one waiting\n");
    }
    BoundedBuffer* newPipe = new BoundedBuffer();
    currentThread->setPipelineWaitingWriter(newPipe);
    newProcess->setWritePipe(newPipe);
  }

  newProcess->Fork(runUserCode, (int)threadArguments);

  return addressSpaceId;
}
