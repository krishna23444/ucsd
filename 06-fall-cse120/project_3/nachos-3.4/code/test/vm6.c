/*
 * Program that demonstrates Random locality
 * 
 * Expected output:
 * Ticks: total 13598891, idle 0, system 4486690, user 9112201
 * Disk I/O: reads 0, writes 0
 * Console I/O: reads 0, writes 0
 * Paging: faults 112165, out 98298
 * Network I/O: packets received 0, sent 0
 * Physical memory size: 32 pages
 * 
 * Cleaning up...
 * 
 */
 #include "syscall.h"
 
unsigned int seed;
int arr[100000];
void
irand_init()
{
  seed = 93186752;
}

/* a very poor random number generator */
unsigned int
irand(int l, int h)
{
  unsigned int a = 1588635695, m = 429496U, q = 2, r = 1117695901;
  unsigned int val;

  seed = a*(seed % q) - r*(seed / q);
  val = (seed / m) % (h - l) + l;
 
  return val;
}

int main() {
  int arr_size = sizeof(arr)/sizeof(arr[0]);
  int i;
  
  irand_init();
  for (i = 0; i < arr_size; i++) {
    arr[irand(0,arr_size)] = i;
  }
  Exit(0);
}
