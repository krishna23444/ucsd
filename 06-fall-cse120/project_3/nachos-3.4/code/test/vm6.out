Thread spaceID 1 Exiting with value 0
No threads ready or runnable, and no pending interrupts.
Assuming the program completed.
Machine halting!

Ticks: total 13598891, idle 0, system 4486690, user 9112201
Disk I/O: reads 0, writes 0
Console I/O: reads 0, writes 0
Paging: faults 112165, in 111851, out 98298
Network I/O: packets received 0, sent 0
Physical memory size: 32 pages

Cleaning up...
